### ISSUE TITLE
[Question]:  when break complex problem into sub-problems raise “JSONDecodeError"

### ISSUE THREAD

**Issue by xiaohuozilalala** [run-llama/llama_index#14152]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

When I followed the official documentation guide "[Llama3 Cookbook with Ollama](https://docs.llamaindex.ai/en/stable/examples/cookbooks/llama3_cookbook_ollama_replicate/?h=history#6-adding-chat-history-to-rag-chat-engine)", I encountered a problem in the example "Break Complex Questions down into Sub-Questions". When I tried to run the code according to the tutorial in the example, the compiler prompted an error in JSON parsing, which did not occur in the previous tutorials.

**code show as below:**
```
drake_index = VectorStoreIndex.from_documents(drake_docs)
drake_query_engine = drake_index.as_query_engine(similarity_top_k=3)

kendrick_index = VectorStoreIndex.from_documents(kendrick_docs)
kendrick_query_engine = kendrick_index.as_query_engine(similarity_top_k=3)

drake_tool = QueryEngineTool(
    drake_index.as_query_engine(),
    metadata=ToolMetadata(
        name="drake_search",
        description="Useful for searching over Drake's life.",
    ),
)

kendrick_tool = QueryEngineTool(
    kendrick_index.as_query_engine(),
    metadata=ToolMetadata(
        name="kendrick_search",
        description="Useful for searching over Kendrick's life.",
    ),
)

query_engine = SubQuestionQueryEngine.from_defaults(
    [drake_tool, kendrick_tool],
    llm=llm,  # llama3-8b,Ollama
    verbose=True,
)

response = query_engine.query("Which albums did Drake release in his career?")

print(response)
```

**The compiler reports the following error:**
```
Traceback (most recent call last):
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\output_parsers\utils.py", line 45, in parse_json_markdown
    json_obj = json.loads(json_string)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\json\decoder.py", line 340, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 21 column 1 (char 500)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\output_parsers\utils.py", line 52, in parse_json_markdown
    json_obj = yaml.safe_load(json_string)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\__init__.py", line 81, in load
    return loader.get_single_data()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\constructor.py", line 49, in get_single_data
    node = self.get_single_node()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\composer.py", line 58, in compose_document
    self.get_event()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\parser.py", line 118, in get_event
    self.current_event = self.state()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\parser.py", line 193, in parse_document_end
    token = self.peek_token()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\scanner.py", line 129, in peek_token
    self.fetch_more_tokens()
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\yaml\scanner.py", line 258, in fetch_more_tokens
    raise ScannerError("while scanning for the next token", None,
yaml.scanner.ScannerError: while scanning for the next token
found character '`' that cannot start any token
  in "<unicode string>", line 21, column 1:
    ```
    ^

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\CodeProject\langchain1\pythonProject1\ollama-langchain-example.py", line 191, in <module>
    response = query_engine.query("Which albums did Drake release in his career?")
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\instrumentation\dispatcher.py", line 198, in wrapper
    result = func(*args, **kwargs)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\base\base_query_engine.py", line 51, in query
    query_result = self._query(str_or_query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\query_engine\sub_question_query_engine.py", line 145, in _query
    sub_questions = self._question_gen.generate(self._metadatas, query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\llm_generators.py", line 81, in generate
    parse = self._prompt.output_parser.parse(prediction)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\output_parser.py", line 11, in parse
    json_dict = parse_json_markdown(output)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\output_parsers\utils.py", line 54, in parse_json_markdown
    raise OutputParserException(
llama_index.core.output_parsers.base.OutputParserException: Got invalid JSON object. Error: Extra data: line 21 column 1 (char 500) while scanning for the next token
found character '`' that cannot start any token
  in "<unicode string>", line 21, column 1:
    ```
    ^. Got JSON string: {
    "items": [
        {
            "sub_question": "What is the revenue growth of Uber",
            "tool_name": "uber_10k"
        },
        {
            "sub_question": "What is the EBITDA of Uber",
            "tool_name": "uber_10k"
        },
        {
            "sub_question": "What is the revenue growth of Lyft",
            "tool_name": "lyft_10k"
        },
        {
            "sub_question": "What is the EBITDA of Lyft",
            "tool_name": "lyft_10k"
        }
    ]
}

# Example 2

{
    "items": [
        {
            "sub_question": "What are Drake's albums?",
            "tool_name": "drake_search"
        }
    ]
}

Process finished with exit code 1
```

In the error message, there is an error message about JSON parsing. What could be the reason? . What also puzzles me is that the error message contains information about previously used tools (such as "uber_10k", "lyft_10k", etc.), which seem to be loaded, but I did not call them. Is there any solution? Thank you very much!

**Comment by xiaohuozilalala** [comment]
@dosu Thank you for the ideas you provided. But after checking and running the code, I found that the JSON parsing error seems to be a superficial reason, not the most fundamental reason. The "query_engine()" method seems to automatically cache my previous questions and keep them in memory. As a result, in the current task, even if I did not mention the previous questions, they were added to the current questions by default, causing errors. Because the compiler's error repeatedly mentioned information related to previous Uber and Lyft financial reports, and the document was not used in the current task. Is this a memory leak, or is it caused by automatic caching in query_engine()? Besides that, this JSON error problem seems to be related to the decomposition of sub-problems. Could it be that "query_engine()" fails to decompose the sub-problems correctly, causing the error?

Here are the new errors the compiler reports (which keep changing even if the code is not modified):
error1
```
Traceback (most recent call last):
  File "E:\CodeProject\langchain1\pythonProject1\ollama-langchain-example.py", line 190, in <module>
    response = query_engine.query("Which albums did Drake release in his career?")
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\instrumentation\dispatcher.py", line 198, in wrapper
    result = func(*args, **kwargs)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\base\base_query_engine.py", line 51, in query
    query_result = self._query(str_or_query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\query_engine\sub_question_query_engine.py", line 145, in _query
    sub_questions = self._question_gen.generate(self._metadatas, query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\llm_generators.py", line 81, in generate
    parse = self._prompt.output_parser.parse(prediction)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\output_parser.py", line 13, in parse
    raise ValueError(f"No valid JSON found in output: {output}")
ValueError: No valid JSON found in output: I'd be happy to help! Given a user question and a list of tools, I can output a list of relevant sub-questions in JSON markdown that when composed can help answer the full user question.

Please provide the user question and the list of tools, and I'll generate the output. For example 1, I understand that you want me to compare and contrast revenue growth and EBITDA of Uber and Lyft for year 2021.
```
error2
```
Traceback (most recent call last):
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\pydantic\v1\main.py", line 522, in parse_obj
    obj = dict(obj)
ValueError: dictionary update sequence element #0 has length 1; 2 is required

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "E:\CodeProject\langchain1\pythonProject1\ollama-langchain-example.py", line 190, in <module>
    response = query_engine.query("Which albums did Drake release in his career?")
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\instrumentation\dispatcher.py", line 198, in wrapper
    result = func(*args, **kwargs)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\base\base_query_engine.py", line 51, in query
    query_result = self._query(str_or_query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\query_engine\sub_question_query_engine.py", line 145, in _query
    sub_questions = self._question_gen.generate(self._metadatas, query_bundle)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\llm_generators.py", line 81, in generate
    parse = self._prompt.output_parser.parse(prediction)
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\output_parser.py", line 21, in parse
    sub_questions = [SubQuestion.parse_obj(item) for item in json_dict]
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\llama_index\core\question_gen\output_parser.py", line 21, in <listcomp>
    sub_questions = [SubQuestion.parse_obj(item) for item in json_dict]
  File "E:\CodeProject\conda_file\envs\langchain1\lib\site-packages\pydantic\v1\main.py", line 525, in parse_obj
    raise ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls) from e
pydantic.v1.error_wrappers.ValidationError: 1 validation error for SubQuestion
__root__
  SubQuestion expected dict not str (type=type_error)
```

The error message seems to be different each time I run it, which makes me very confused. How can I further investigate?Thanks!

**Comment by xiaohuozilalala** [comment]
@dosu Thank you for your reply. After rechecking, I found that this problem may be similar to #7582 and  #8701. This seems to be related to the llm engine used: when using "gpt-3.5-turbo", the code can work properly and give reasonable answers. Then I tested different llm models in Ollama and found that when using "phi3-14b" and "qwen2-7b", the code can run, but the responses given are not reasonable; and when using "llama3-8b", the code directly reports an error and cannot run. So, if I want to use an open source model with smaller parameters such as "llama3-8b" and combine it with "query_engine" to decompose sub-problems in RAG, how can I avoid this problem?
